src/library/dao/playlistdao.cpp

```diff
@@ -723,54 +751,11 @@ int PlaylistDAO::getMaxPosition(const int playlistId) const {
 }
 
 void PlaylistDAO::removeTracksFromPlaylists(const QList<TrackId>& trackIds) {
-    QStringList trackIdList;
-    for (const auto& trackId: trackIds) {
-        if (trackIdList.count() >= 255) {
-            // Avoid that the resulting SQL query to exceed the maximum length
-            // The maximum number of bytes in the text of an SQL statement is
-            // limited to SQLITE_MAX_SQL_LENGTH which defaults to 1000000
-            // (from http://www.sqlite.org/limits.html)
-            removeTracksFromPlaylistsInner(trackIdList);
-            trackIdList.clear();
-        }
-        trackIdList << trackId.toString();
-    }
-    removeTracksFromPlaylistsInner(trackIdList);
-
-    // Update the in-memory hash. TODO(XXX) this assumes all the removes
-    // succeeded.
     for (const auto& trackId: trackIds) {
-        m_playlistsTrackIsIn.remove(trackId);
-    }
-
-}
-
-void PlaylistDAO::removeTracksFromPlaylistsInner(const QStringList& trackIdList) {
-    QSqlQuery query(m_database);
-    query.prepare(QString("SELECT DISTINCT playlist_id FROM PlaylistTracks WHERE track_id in (%1)")
-                  .arg(trackIdList.join(",")));
-
-    if (!query.exec()) {
-        LOG_FAILED_QUERY(query);
-        return;
-    }
-
-    // Collect all ids of the playlists that contains the tracks to remove
-    QList<int> removedTracksPlaylistIds;
-    const int playlistIDColoumn = query.record().indexOf("playlist_id");
-    while (query.next()) {
-        removedTracksPlaylistIds.append(query.value(playlistIDColoumn).toInt());
-    }
-
-    query.prepare(QString("DELETE FROM PlaylistTracks WHERE track_id in (%1)")
-                  .arg(trackIdList.join(",")));
-    if (!query.exec()) {
-        LOG_FAILED_QUERY(query);
-        return;
-    }
-
-    foreach (int playlistId, removedTracksPlaylistIds) {
-        emit(changed(playlistId));
+    	const auto it = m_playlistsTrackIsIn.find(trackId);
+    	while (it != m_playlistsTrackIsIn.end() && it.key() == trackId) {
+    		removeTrackFromPlaylist(it.value(), trackId);
```

Comment:
The code does not reveal why it is safe to keep iterating while m_playlistsTrackIsIn is modified in removeTrackFromPlaylist(). Please check that this is true and add a comment. Or take the safe route and store the result of the iteration in a temporary collection. I would recommend the latter.